pkg/util/labels.go

```diff
@@ -7,43 +7,61 @@ import (
+	var buildNamespace string
+	if len(config.BuildNamespaceLabel) > 0 {
+		buildNamespace = config.BuildNamespaceLabel
+	} else {
+		buildNamespace = api.DefaultNamespace
+	}
+
+	labels = GenerateLabelsFromConfig(labels, config, buildNamespace)
+	labels = GenerateLabelsFromSourceInfo(labels, info, buildNamespace)
+	return labels
+}
 
+// GenerateLabelsFromConfig generate the labels based on build s2i Config
+func GenerateLabelsFromConfig(labels map[string]string, config *api.Config, buildNamespace string) map[string]string {
 	if len(config.Description) > 0 {
-		result[api.KubernetesNamespace+"description"] = config.Description
+		labels[api.KubernetesNamespace+"description"] = config.Description
 	}
 
 	if len(config.DisplayName) > 0 {
-		result[api.KubernetesNamespace+"display-name"] = config.DisplayName
+		labels[api.KubernetesNamespace+"display-name"] = config.DisplayName
 	} else {
-		result[api.KubernetesNamespace+"display-name"] = config.Tag
+		labels[api.KubernetesNamespace+"display-name"] = config.Tag
 	}
 
+	addBuildLabel(labels, "image", config.BuilderImage, buildNamespace)
+	addBuildLabel(labels, "source-context-dir", config.ContextDir, buildNamespace)
+	return labels
+}
+
+// GenerateLabelsFromSourceInfo generate the labels based on the source repository
+// informations.
+func GenerateLabelsFromSourceInfo(labels map[string]string, info *api.SourceInfo, buildNamespace string) map[string]string {
 	if info == nil {
 		glog.V(3).Infof("Unable to fetch source informations, the output image labels will not be set")
-		return result
+		return labels
 	}
 
-	addBuildLabel(result, "image", config.BuilderImage)
-	addBuildLabel(result, "commit.author", info.Author)
-	addBuildLabel(result, "commit.date", info.Date)
-	addBuildLabel(result, "commit.id", info.CommitID)
-	addBuildLabel(result, "commit.ref", info.Ref)
-	addBuildLabel(result, "commit.message", info.Message)
-	addBuildLabel(result, "source-location", info.Location)
-	addBuildLabel(result, "source-context-dir", config.ContextDir)
-
-	return result
+	addBuildLabel(labels, "commit.author", info.Author, buildNamespace)
+	addBuildLabel(labels, "commit.date", info.Date, buildNamespace)
+	addBuildLabel(labels, "commit.id", info.CommitID, buildNamespace)
+	addBuildLabel(labels, "commit.ref", info.Ref, buildNamespace)
+	addBuildLabel(labels, "commit.message", info.Message, buildNamespace)
+	addBuildLabel(labels, "source-location", info.Location, buildNamespace)
+	return labels
 }
 
-// addBuildLabel adds a new "io.openshift.s2i.build.*" label into map when the
+// addBuildLabel adds a new "*.build.*" label into map when the
 // value of this label is not empty
-func addBuildLabel(to map[string]string, key, value string) {
+func addBuildLabel(to map[string]string, key, value, buildNamespace string) {
 	if len(value) == 0 {
 		return
 	}
-	to[strings.Join([]string{api.DefaultNamespace, "build.", key}, "")] = value
+	to[strings.Join([]string{buildNamespace, "build.", key}, "")] = value
```

Comment:
@rhcarvalho http://herman.asia/efficient-string-concatenation-in-go, O(n2) is slower than O(n) imho. In your tests you're just testing times, but not allocations, which makes the difference in the long run. Remember strings in go are immutable, similarly to python.